---
title: "Differential expression analysis of individual Drosophila melanogaster (w1118) male and female flies fed with different concentrations of dimethyl sulfoxide (DMSO) over different time periods."
author: "Analysis done by Yuejun Wang, TriLab Biounformatic Group"
output:
  html_document:
  pdf_document: default
  df_print: paged
gemoetry: margin=1in
fontsize: 8pt
fontfamily: mathpazo
---

\center

# *Summary:*
To gain an understanding of the toxic effect of a commonly used solvent, flies were exposed to 0, 0.5, 2, 3.5, or 7% v/v DMSO for 2, 4, 8, 24 or 48 h. We performed compound exposure of 800 individual flies in 4 Whole Animal Feeding Flats (WAFFL), a novel 96 well system to house, feed, and harvest individual flies. This expression profiling was part of a set of the experiments performed to evaluate the suitability of the WAFFL for high throughput small compound screening in D. melanogaster. Treated flies and controls were used for poly A+ stranded mRNA library preparation and we performed high throughput RNA sequencing to determine the transcriptional changes due to DMSO treatment.

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, tidy = TRUE)
```

# Load libraries
```{r message=FALSE}

options(java.parameters = "-Xmx8000m")

library(tidyverse)
library(DESeq2)
library(cowplot)
library(ggpubr)
library(RColorBrewer)
library(pheatmap)
library(ggsci)
library(AnnotationDbi)
library(scales)
library(org.Dm.eg.db)
library(xlsx)
library(r2excel)
library(ggplot2)
library(viridis)
library(RSkittleBrewer)
library(genefilter)
library(corrplot)
library(reshape2)
library(ComplexHeatmap)
library(circlize)
library(UpSetR)
trop = RSkittleBrewer('tropical')
```

# Load aux functions
```{r}
source(file = "01_aux_rnaseq_functions.R")
```

# Load data
```{r}
metadata <- as.data.frame(read_tsv(file = "SampleInfo.txt", col_names = TRUE, comment = "#"))
rownames(metadata) <- metadata$SampleName


read_counts <- as.data.frame(read_tsv(file = "/gpfs/gsfs12/users/wangy80/TK74/results/05counts/read_counts", col_names = TRUE, comment = "#"))
rownames(read_counts) <- read_counts$Geneid # adding gene ids as row names


# Rename sample names
read_counts <- read_counts[,7:length(read_counts)]

colnames(read_counts) <- gsub(".sorted.dedup.bam","",
                              gsub("results/04dedup/","",
                                   colnames(read_counts)))


# Sort tables so metadata and read counts match order
read_counts <- read_counts[,match(metadata$SampleName, colnames(read_counts))]
all(rownames(metadata) == colnames(read_counts))


# Round read counts to the closest interger
read_counts <- round(read_counts, digits = 0)

# include total read counts in metadata
metadata$read_counts <- colSums(read_counts)

# Rename column names in read_counts based on metadata
colnames(read_counts) <- rownames(metadata)

write.table(x = metadata, file = "metadata.txt", sep = "\t") 
```

# DE analysis with DESeq2

```{r}
dir.create(path = "./Plots", showWarnings = FALSE)

# Converting Time, DMSO and Replicate number to factors
metadata$Time <- factor(metadata$Time,levels=c("2hr","4hr","8hr","24hr","48hr"))
metadata$DMSO <- as.factor(metadata$DMSO)
metadata$Replicate <- as.factor(metadata$Replicate)
metadata$Sex <- as.factor(metadata$Sex)

# Adding read_depth in design to control for read_depth
dds <- DESeqDataSetFromMatrix(countData = read_counts, 
                              colData = metadata,  
                              design = ~ Sex + Time + DMSO)


# Plot total reads per sample using barchar
p <- ggbarplot(data = metadata, 
          x = "SampleName", 
          y = "read_counts",
          x.text.angle = 90,
          fill = "DMSO", 
          title = "Total read counts per sample", 
          ylab = "Read counts",
          sort.by.groups = TRUE,
          palette = "jco",
          sort.val = "asc")

ggsave2("Plots/barplot_read_counts_per_sample.pdf", plot = p,width=14,height=8)

# Normalize counts
vsd <- vst(dds, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(as.data.frame(dds@assays@data@listData)) >= 20
vsd <- vsd[keep,]

# Calculate distances between samples
sampleDists <- dist(t(assay(vsd)))

# Plot inter-sample distances
old.par <- par(no.readonly=T)

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$SampleName)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
p.hm <- pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         color = colorRampPalette(c("red", "white", "blue"))(100))

pdf("./Plots/heat_map.pdf", p.hm,width=16,height=16)
p.hm
dev.off()

p.hm

# PCA
pcaData <- plotPCA(vsd, intgroup=c("DMSO","Time","Sex"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
y.coords = c(min(pcaData$PC1, pcaData$PC2), max(pcaData$PC1, pcaData$PC2))
x.coords = y.coords
p1 <- ggplot(pcaData, aes(PC1, PC2, color=DMSO, shape=Sex, size=Time)) +
  geom_point() + scale_color_lancet() + 
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData$PC1)-min(pcaData$PC1))/(max(pcaData$PC2)-min(pcaData$PC2))) 

ggsave("Plots/pca.pdf", plot = p1)
p1

```

Samples clustered mainly by Sex.

## pearson correlation

```{r}
df <- as.data.frame(assay(vsd))
write.table(df,"DESeq2_normalized.txt",sep="\t",quote=F)
correlation <- cor(df)

# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}


reorder_cormat <- function(cormat){
  # Use correlation between variables as distance
  dd <- as.dist((1-cormat)/2)
  hc <- hclust(dd)
  cormat <-cormat[hc$order, hc$order]
}

# Reorder the correlation matrix
cormat <- reorder_cormat(correlation)
upper_tri <- get_upper_tri(cormat)
# Melt the correlation matrix
melted_cormat <- melt(upper_tri, na.rm = TRUE)
# Create a ggheatmap
ggheatmap <- ggplot(melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0.75, limit = c(0.5,1), space = "Lab", 
                       name="Pearson\nCorrelation") +
  theme_minimal()+ # minimal theme
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, 
                                   size = 12, hjust = 1))+
  coord_fixed()
# Print the heatmap
print(ggheatmap)


pdf("./Plots/DESeq2_sample_pearson_correlation.pdf",height=14,width=15)
ggheatmap
dev.off()
```

```{r}
marker_gene <- c("lncRNA:roX2","lncRNA:roX1","Yp2","Yp1","Yp3")

# Add gene symbols
symbol_list <- replace_gene_acc_by_symbol_ids(rownames(df), return_all = TRUE, db = org.Dm.eg.db)
df$Gene_name <- symbol_list

res <- df[df$Gene_name %in% marker_gene,]
rownames(res) <- res$Gene_name

res2 <- t(scale(t(res[,1:ncol(res)-1])))

p1 <- Heatmap(res2, 
      name="zscore",
      show_row_names = TRUE
      #col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
      )

pdf("./Plots/DESeq2_markers_expression_zscore.pdf",height=5,width=18)
p1
dev.off()

p1


```

# Separate M and F

```{r}
dds.f <- dds[, dds$Sex == "F"]
dds.f$Sex <- droplevels(dds.f$Sex)
dds.f$DMSO <- relevel(dds.f$DMSO, ref = "0")
#dds.f$Time <- relevel(dds.f$Time, ref = "2hr")
dds.f$group <- factor(paste0(dds.f$Time,"_DMSO",dds.f$DMSO))

design(dds.f) <- ~group

dds.f <- DESeq(dds.f)
dds.f$dds_ID = "female"
resultsNames(dds.f)


#  [1] "Intercept"                        "group_24hr_DMSO0.5_vs_24hr_DMSO0"
#  [3] "group_24hr_DMSO2_vs_24hr_DMSO0"   "group_24hr_DMSO3.5_vs_24hr_DMSO0"
#  [5] "group_24hr_DMSO7_vs_24hr_DMSO0"   "group_2hr_DMSO0_vs_24hr_DMSO0"   
#  [7] "group_2hr_DMSO0.5_vs_24hr_DMSO0"  "group_48hr_DMSO0_vs_24hr_DMSO0"  
#  [9] "group_48hr_DMSO0.5_vs_24hr_DMSO0" "group_48hr_DMSO2_vs_24hr_DMSO0"  
# [11] "group_48hr_DMSO3.5_vs_24hr_DMSO0" "group_48hr_DMSO7_vs_24hr_DMSO0"  
# [13] "group_4hr_DMSO0_vs_24hr_DMSO0"    "group_4hr_DMSO0.5_vs_24hr_DMSO0" 
# [15] "group_8hr_DMSO0_vs_24hr_DMSO0"    "group_8hr_DMSO0.5_vs_24hr_DMSO0" 

dds.m <- dds[,dds$Sex == "M" & dds$SampleName != "M_0_24hr_rep1"] # I removed M_0_24hr_rep1 because it is clustered separated from the rest of the Male samples.
dds.m$Sex <- droplevels(dds.m$Sex)
dds.m$DMSO <- relevel(dds.m$DMSO, "0")
#dds.m$Time <- relevel(dds.m$Time, "2hr")
dds.m$group <- factor(paste0(dds.m$Time,"_DMSO",dds.m$DMSO))

design(dds.m) <- ~group
dds.m <- DESeq(dds.m)
dds.m$dds_ID = "male"
resultsNames(dds.m)

#  [1] "Intercept"                        "group_24hr_DMSO0.5_vs_24hr_DMSO0"
#  [3] "group_24hr_DMSO2_vs_24hr_DMSO0"   "group_24hr_DMSO3.5_vs_24hr_DMSO0"
#  [5] "group_24hr_DMSO7_vs_24hr_DMSO0"   "group_2hr_DMSO0_vs_24hr_DMSO0"   
#  [7] "group_2hr_DMSO0.5_vs_24hr_DMSO0"  "group_48hr_DMSO0_vs_24hr_DMSO0"  
#  [9] "group_48hr_DMSO0.5_vs_24hr_DMSO0" "group_48hr_DMSO2_vs_24hr_DMSO0"  
# [11] "group_48hr_DMSO3.5_vs_24hr_DMSO0" "group_48hr_DMSO7_vs_24hr_DMSO0"  
# [13] "group_4hr_DMSO0_vs_24hr_DMSO0"    "group_4hr_DMSO0.5_vs_24hr_DMSO0" 
# [15] "group_8hr_DMSO0_vs_24hr_DMSO0"    "group_8hr_DMSO0.5_vs_24hr_DMSO0" 

```


# Perform Sex-specific PCA analysis

```{r}
# Females
# Normalize counts
vsd.f <- vst(dds.f, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(as.data.frame(dds.f@assays@data@listData),na.rm = TRUE) >= 20
vsd.f <- vsd.f[keep,]

# Calculate distances between samples
sampleDists.f <- dist(t(assay(vsd.f)))

# Define the number of colors you want
colourCount = length(unique(vsd.f$group))
getPalette = colorRampPalette(brewer.pal(12, "Paired"))

# PCA
pcaData.f <- DESeq2::plotPCA(vsd.f, intgroup=c("group"), returnData=TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData.f, "percentVar"))
y.coords = c(min(pcaData.f$PC1, pcaData.f$PC2), max(pcaData.f$PC1, pcaData.f$PC2))
x.coords = y.coords
p1 <- ggplot(pcaData.f, aes(PC1, PC2, color=group)) +
  geom_point(size=4) + 
  scale_color_manual(values = getPalette(colourCount)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData.f$PC1)-min(pcaData.f$PC1))/(max(pcaData.f$PC2)-min(pcaData.f$PC2))) + labs(title = "Females")

ggsave("Plots/pca_by_female.pdf", plot = p1)
p1

## MALES ## 
# Normalize counts
vsd.m <- vst(dds.m, blind=FALSE)

# Keep genes with at least 20 reads total across samples
keep <- rowSums(as.data.frame(dds.m@assays@data@listData),na.rm = TRUE) >= 20
vsd.m <- vsd.m[keep,]

# Calculate distances between samples
sampleDists.m <- dist(t(assay(vsd.m)))

# PCA
pcaData.m <- DESeq2::plotPCA(vsd.m, intgroup=c("group"), returnData=TRUE, ntop = 500)
percentVar <- round(100 * attr(pcaData.m, "percentVar"))
y.coords = c(min(pcaData.m$PC1, pcaData.m$PC2), max(pcaData.m$PC1, pcaData.m$PC2))
x.coords = y.coords
p2 <- ggplot(pcaData.m, aes(PC1, PC2, color=group)) +
  geom_point(size=4) + 
  scale_color_manual(values = getPalette(colourCount)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed(ratio = (max(pcaData.m$PC1)-min(pcaData.m$PC1))/(max(pcaData.m$PC2)-min(pcaData.m$PC2))) + labs(title = "Males")

ggsave("Plots/pca_by_male.pdf", plot = p2)
p2
```


# Run DE analysis
```{r}

dir.create(path = "./DE", showWarnings = FALSE)

comparisons <- list(
c("group","24hr_DMSO2","24hr_DMSO0"),
c("group","48hr_DMSO0.5","48hr_DMSO0"),
c("group","2hr_DMSO0.5","2hr_DMSO0"),
c("group","8hr_DMSO0.5","8hr_DMSO0"),
c("group","48hr_DMSO2","48hr_DMSO0"),
c("group","24hr_DMSO3.5","24hr_DMSO0"),
c("group","48hr_DMSO3.5","48hr_DMSO0"),
c("group","24hr_DMSO7","24hr_DMSO0"),
c("group","48hr_DMSO7","48hr_DMSO0"),
c("group","4hr_DMSO0.5","4hr_DMSO0"),
c("group","24hr_DMSO0.5","24hr_DMSO0")
)


########################################################################################################
# Make function here to compute DE analysis and output results for each comparison above.
########################################################################################################
# Define function for processing and saving result tables
sort_and_write_res_table <- function(result_table, 
                                     file_name = "TABLE.xlsx", 
                                     sheet_name = "sheet_1"){
  dir.create(path = "./DE", showWarnings = FALSE)
  
  # add path
  file_name <- paste0("./DE/",file_name)
  
  # Check if an excel spreadsheet already exists, otherwise create one
  if(file.exists(file_name)){
    wb <- loadWorkbook(file = file_name)
  } else {
    wb <- createWorkbook(type="xlsx")
  }
  
  # # Create new excel sheet, remove sheets if it already exits (if the sheet name is too long, the errors might occur)
  sheets <- getSheets(wb)
  if(is.element(sheet_name,names(sheets))){
    removeSheet(wb, sheetName=sheet_name)
  }
  sheet <- createSheet(wb, sheetName = sheet_name)

  # Sort genes by (padj)
  result_table_sorted <- result_table[order(result_table$padj, decreasing = FALSE),]
  
  # Add gene symbols
  symbol_list <- replace_gene_acc_by_symbol_ids(rownames(result_table_sorted), return_all = TRUE, db = org.Dm.eg.db)
  result_table_sorted$Gene_name <- symbol_list
  result_table_sorted <- as.data.frame(result_table_sorted) 
  
  # Add df to excel sheet
  xlsx.addTable(wb = wb, sheet = sheet, data = result_table_sorted, startRow = 1, startCol = 1)

  # Write sorted table to Excel file as different worksheets. Need file name + Worksheet name !!!
  saveWorkbook(wb, file_name)
  
  # Write sorted table to file
  # write.table(as.data.frame(result_table_sorted), file = paste0("./DE/",file_name,".txt"), sep = "\t", col.names=NA)
  
  return(result_table_sorted)
}


get_deseq_result <- function(dds, contrast, analysis_type = "ashr"){
  #print(1)
  if(analysis_type %in% c("ashr","apeglm","normal")){
    res <- lfcShrink(dds = dds, contrast =list(contrast), type = analysis_type)
  } else if (analysis_type == "result"){
    res <- results(object = dds, contrast = contrast, independentFiltering = FALSE)
  }
  #print(2)
  # Replace NAs by 1s
  res$pvalue[is.na(res$pvalue)] <- 1
  res$padj[is.na(res$padj)] <- 1
  #print(3)
  # Print out summary of results
  print(contrast)
  summary(res, alpha = 0.05)
  #print(4)
  # Sort result table and save it
  my_file_name = paste0("DE_",dds$dds_ID[1])
  #paste0("DE_",paste0(contrast, collapse = "_"),"_",dds$dds_ID[1])
  
  my_sheet = paste0(paste0(contrast, collapse = "_"),"_",dds$dds_ID[1])
  my_sheet = gsub("group_","",my_sheet)
  #print(5)
  res_sorted <- sort_and_write_res_table(result_table = res, 
                                         file_name =  paste0(my_file_name,".xlsx"), 
                                         sheet_name = my_sheet)
  #print(6)
  return(res_sorted)
  
}
```

```{r}
# Using lfcShrink instead of results to reduce high Log2FC bias of genes with low expression

DE_results.f = list()
DE_results.m = list()
for (i in 1:length(comparisons)){
  print(i)
  id <- paste0(comparisons[[i]], collapse = "_")
  DE_results.f[[id]] <- get_deseq_result(dds = dds.f, contrast = comparisons[[i]], analysis_type = "result" )
  DE_results.m[[id]] <- get_deseq_result(dds = dds.m, contrast = comparisons[[i]], analysis_type = "result" )

  genes_of_interest.f <- head(rownames(DE_results.f[[id]][DE_results.f[[id]]$padj <= 0.05, ]))
  if (length(genes_of_interest.f) > 0){
    plot_normalized_gene_expression(my_dds = dds.f, 
                                  ensmbl_gene_list = genes_of_interest.f, 
                                  file_prefix = paste0("Gene_exp_",id, "_F_")) 
  }
  genes_of_interest.m <- head(rownames(DE_results.m[[id]][DE_results.m[[id]]$padj <= 0.05, ]))
  if (length(genes_of_interest.m) > 0){
    plot_normalized_gene_expression(my_dds = dds.m, 
                                  ensmbl_gene_list = genes_of_interest.m, 
                                  file_prefix = paste0("Gene_exp_",id, "_M_")) 
  }
}  

```

# genelists overlap

```{r}
my_overlap_list.f <- list()
for (i in names(DE_results.f)){
    DE <- c(rownames(subset(DE_results.f[[i]], padj <= 0.05)))
    my_overlap_list.f[[i]] <- DE
}


sub1 <- my_overlap_list.f[grep("DMSO0.5", names(my_overlap_list.f))]
sub2 <- my_overlap_list.f[-grep("DMSO0.5", names(my_overlap_list.f))]


m1 = make_comb_mat(sub1)
p1 <- UpSet(m1,top_annotation = upset_top_annotation(m1, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m1, add_numbers = TRUE))

m2 = make_comb_mat(sub2)
p2 <- UpSet(m2,top_annotation = upset_top_annotation(m2, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m2, add_numbers = TRUE))

pdf("./Plots/overlap_female_DEs_sub1.pdf",height=5,width=10)
p1
dev.off()

pdf("./Plots/overlap_female_DEs_sub2.pdf",height=8,width=18)
p2
dev.off()

# male
my_overlap_list.m <- list()
for (i in names(DE_results.m)){
    DE <- c(rownames(subset(DE_results.m[[i]], padj <= 0.05)))
    my_overlap_list.m[[i]] <- DE
}


sub1 <- my_overlap_list.m[grep("DMSO0.5", names(my_overlap_list.m))]
sub2 <- my_overlap_list.m[-grep("DMSO0.5", names(my_overlap_list.m))]


m1 = make_comb_mat(sub1)
p1 <- UpSet(m1,top_annotation = upset_top_annotation(m1, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m1, add_numbers = TRUE))

m2 = make_comb_mat(sub2)
p2 <- UpSet(m2,top_annotation = upset_top_annotation(m2, add_numbers = TRUE),
    right_annotation = upset_right_annotation(m2, add_numbers = TRUE))

pdf("./Plots/overlap_male_DEs_sub1.pdf",height=5,width=10)
p1
dev.off()

pdf("./Plots/overlap_male_DEs_sub2.pdf",height=8,width=18)
p2
dev.off()

```



# Build summary table across contrats
```{r}
N <- length(names(DE_results.f))
# Initialize table
my_table <- data.frame(rbind("logFC.up.female" = rep(0, N), 
                             "logFC.down.female" = rep(0, N),
                             "logFC.up.male" = rep(0, N), 
                             "logFC.down.male" = rep(0, N)
                             )
                       ) 
colnames(my_table) <- names(DE_results.f)

for (i in names(DE_results.f)){ 
  my_table["logFC.up.female",i] <- table(DE_results.f[[i]]$padj <= 0.05 & DE_results.f[[i]]$log2FoldChange > 0)[2]
  my_table["logFC.down.female",i] <- table(DE_results.f[[i]]$padj <= 0.05 & DE_results.f[[i]]$log2FoldChange < 0)[2]
  my_table["logFC.up.male",i] <- table(DE_results.m[[i]]$padj <= 0.05 & DE_results.m[[i]]$log2FoldChange > 0)[2]
  my_table["logFC.down.male",i] <- table(DE_results.m[[i]]$padj <= 0.05 & DE_results.m[[i]]$log2FoldChange < 0)[2]
}

names(DE_results.f) <- paste0("female_",names(DE_results.f))
names(DE_results.m) <- paste0("male_",names(DE_results.m))


# Replace NA by 0s
my_table[is.na(my_table )] <- 0
write.table(x = my_table, file = "DE_summary.txt", sep = "\t", col.names = NA)

```


# Look for DE gene overlaps among time points and direction of change in gene expression
# Build summary table across contrats
```{r}

get_diff_exp_genes_across_conditions_plot <- function(DE_result_list){
  # Initialize list 
  my_overlap_list.f <- list()
  all_sig_genes <- c()
  
  for (i in names(DE_result_list)){
    genes_up.f <- c(rownames(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange > 0.5)))
    genes_dn.f <- c(rownames(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange < -0.5)))
    #my_overlap_list.f[i] <- c(subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange > 0)[,"Gene_name"])
    my_overlap_list.f[[i]] <- unique(c(genes_up.f, genes_dn.f) )
    #                          subset(DE_result_list[[i]], padj <= 0.05 & log2FoldChange < 0)[,"Gene_name"]
    #                          )
    all_sig_genes <- unique(c(all_sig_genes, my_overlap_list.f[[i]]))
    #my_table_overlap["logFC.up.male",i] <- subset(DE_results.m[[i]], padj <= 0.05 & log2FoldChange > 0)[,"Gene_name"]
    #my_table_overlap["logFC.down.male",i] <- subset(DE_results.m[[i]], padj <= 0.05 & log2FoldChange < 0)[,"Gene_name"]
  }
  
  row.N <- length(all_sig_genes)
  col.N <- length(names(DE_result_list))
  my_tibble <- as_tibble(matrix(nrow = row.N, ncol = col.N + 2))
  colnames(my_tibble) <- c("gene_ids","gene_names", names(DE_result_list))
  my_tibble$gene_ids <- all_sig_genes
  my_tibble$gene_names <- subset(DE_result_list[[i]], rownames(DE_result_list[[i]]) %in%  all_sig_genes)["Gene_name"]
  
  # Load tibble (1 => is differentially expressed in contrast X; 0 => is NOT differentially expressed in contrast X)
  for (my_row in 1:row.N){
    for (my_col in 1:col.N){
      my_query <- my_tibble$gene_ids[my_row]
      my_contrast <- colnames(my_tibble)[my_col]
      my_tibble[my_row, my_col + 2] <- ifelse(test = my_query %in% my_overlap_list.f[[my_contrast]], 
                                          yes = subset(DE_result_list[[my_contrast]], rownames(DE_result_list[[my_contrast]]) == my_query)["log2FoldChange"], 
                                          no = 0)
    }
  }
  
  my_tibble_long <- pivot_longer(my_tibble, cols = 3:ncol(my_tibble), names_to = "contrast", values_to = "log2fc")
  my_tibble_long$log2fc[my_tibble_long$log2fc < -5] <- -5
  my_tibble_long$log2fc[my_tibble_long$log2fc >  5] <-  5
  
  # Plot table
  p <- my_tibble_long %>% ggplot(aes(x = gene_names$Gene_name, y = contrast, colour = "white", fill = log2fc)) + 
    theme_pubclean() + 
    geom_tile( ) +
    scale_fill_gradient2(low = "red", high = "green", midpoint = 0, mid = "white") + 
    coord_fixed() + 
    theme(axis.ticks = element_blank(),
          panel.grid.minor.x = element_blank(),panel.grid.minor.y = element_blank(),
          panel.grid.major.x = element_blank(),panel.grid.major.y = element_blank(),
          axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
    xlab("Differentially expressed genes") + 
    ylab("Contrasts") + 
    labs(title = "Log2FC of genes differentially expressed across experimental conditions", 
         subtitle = "( | Log2FC | > 0.5 and adj.p < 0.05 )")

    return(p)
}

lists <- c(DE_results.f,DE_results.m)

p <- get_diff_exp_genes_across_conditions_plot(DE_result_list = lists)
ggsave(filename = "perturbed_genes.pdf", plot = p, path = "./Plots", width = 20, height = 10)
p

p.f <- get_diff_exp_genes_across_conditions_plot(DE_result_list = DE_results.f)
ggsave(filename = "perturbed_genes_female.pdf", plot = p.f, path = "./Plots", width = 20, height = 10)
p.f

p.m <- get_diff_exp_genes_across_conditions_plot(DE_result_list = DE_results.m)
ggsave(filename = "perturbed_genes_male.pdf", plot = p.m, path = "./Plots", width = 20, height = 10)
p.m
```

## heatmap for all comparisons on all significant genes detected in at least one comparisons

```{r message=FALSE}

heatmap_DE <- function(list=list,out=out) {
  list<-lapply(list, function(x) 
  cbind(x, gene_ids = rownames(x)))

  filter <- function(df) {
    df[df$padj<=0.05,]
  }

  # Apply the filter function to each data frame in the list
  filtered <- lapply(list, filter)

  # Merge the filtered data frames into a single data frame
  all_sig_genes <- unique((bind_rows(filtered))$gene_ids)

  # Merge data frames by gene_ids
  merged_df <- bind_rows(list, .id = "comparisons")

  filtered_df <- merged_df[merged_df$gene_ids %in% all_sig_genes, ]

  filtered_df_wide <- pivot_wider(filtered_df[,c(1,3,9)],  names_from = "comparisons", values_from = "log2FoldChange")


  # heatmap

  mat <- as.data.frame(filtered_df_wide[,2:ncol(filtered_df_wide)])
  rownames(mat) <- filtered_df_wide$gene_ids

  mat[is.na(mat)] <- 0

  p <- Heatmap(mat, 
        name="log2FoldChange",
        show_row_names = FALSE,
        col = colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
        #column_names_gp = gpar(fontsize =8),
        #heatmap_height = unit(12, "cm"), 
        #heatmap_width = unit(12, "cm")
        )

  pdf(paste0("./Plots/pheatmap_",out,".pdf"),p,height=12,width=12)
  plot(p)
  dev.off()
}

heatmap_DE(list=DE_results.f,out="female_DE")
heatmap_DE(list=DE_results.m,out="male_DE")
heatmap_DE(list=lists,out="all_DE")

```

```{r}
sessionInfo()
```
